### DISCUSSION TITLE
hierarchicalMenu checkbox option

### DISCUSSION THREAD

**Discussion by ryanmitchell** [algolia/instantsearch#5204] [Ideas]
Feature request

Is there any way to have a hierarchicalMenu that shows the lowest level children as a checkbox list rather than a link list, in the same way that a refinement list works? Maybe this is already possible, but I've had a look and a search and can't come across a way of doing it.

Even if you could offer some pointers in how to go about it, I should be able to write the code myself.

Thanks

**Comment by bobylito** [comment]
Hi @ryanmitchell thanks for taking the time to open this issue. Can you give more details about the behaviour that you expect beyond having checkboxes? Would you like to be able to select multiple items in the hierarchy?

**Comment by ryanmitchell** [comment]
Hello

Yes it would need to be able to multi-select. Effectively we are trying to do a refinementList but with a hierarchy, allowing us to have grouped lists for refinement rather than one big list.

**Comment by bobylito** [comment]
Thanks for the details, @ryanmitchell It is not possible as for now.

**Comment by ryanmitchell** [comment]
No problem - if I was to work on it would it be of interest to you as a PR?

**Comment by ryanmitchell** [comment]
@bobylito Just one further question - is there any way of getting the complete hierarchy with the number of results per item? Currently I can only get the main root plus the level i've refined to, but ideally I'd like to be able to get everything. Otherwise its back to the drawing board.

**Comment by vvo** [comment]
Do you mean you would want to show the hierarchical menu with every menu item opened showing the whole tree at start?

**Comment by hereswhatidid** [comment]
I know this is an older ticket but I would like to know if it's possible to do that last suggestion: show the full tree at the beginning rather than having to click down to see the lower levels.

**Comment by Haroenv** [comment]
Hey @hereswhatidid, that information is actually not readily available at initial rendering, because each of the lower levels would have to do another query. 

For that reason an “opened” hierarchicalMenu isn’t simply possible. 

A disjunctive (multiple items checked possible) might be possible however, but as discussed in https://github.com/algolia/instantsearch.js/issues/2053, this needs a lower level solution in the [helper](https://github.com/algolia/algoliasearch-helper-js).

Feel free to give it a try though!

**Comment by hereswhatidid** [comment]
Gotcha, thanks!

**Comment by Haroenv** [comment]
Closing in favour of #2053

**Comment by Haroenv** [comment]
never mind, this is the only issue open on disjunctive hierarchicalMenu (aka hierarchicalRefinementList) for is.js

**Comment by slimtimbodin** [comment]
@ryanmitchell  - Were you able to come up with a way to do this?  Basically I need an instantSearch widget as you described.  One that allows the user to make multiple selections.

For example, if a hierarchy is States and the cities in those states, the user wants to be able to select multiple cities.  The hit container would then display those items that have all the cities ANDed together.

One modification I have been attempting to do is I am trying to see if each time I click on on item in the hierarchicalMenu it would just keep adding to the refinement list instead of the current behaviour where the next one I click removes the previous refinement.

Any advice you or anyone has would be appreciated.

**Comment by ryanmitchell** [comment]
I ended up storing the list of categories in a normal SQL database and then outputting that aspect outside of instantsearch, with an onchange event in JS to update instantsearch results.

**Comment by joshuabaker** [comment]
+1

I’ve ended up here a few times now. This seems to be a search paradigm that we keep coming across.

**Comment by Haroenv** [comment]
@joshuabaker, feel free to upvote the original issue 😄

**Comment by piersb** [comment]
@ryanmitchell I have the same issue - is there any chance that you could share some of your code and save me from reinventing the wheel?

**Comment by ryanmitchell** [comment]
@piersb - nothing to share really. We just output the refinements using SQL and then using an on change event added or removed them from the refinement list. Pretty simple stuff.

**Comment by piersb** [comment]
Ta.

**Comment by sandipbhuyan** [comment]
I have faced the same issue. Basically, In my application, an Entity can have multiple categories. So for searching through these categories, I need to implement multiple category selection features. Is there a way to do this?

**Comment by Haroenv** [comment]
For now there's no way to do this in the Algolia backend. A solution could be found by combining refinement lists to make a new kind of hierarchical refinement instead

**Comment by sandipbhuyan** [comment]
One problem I am facing here. I have indexed the category details in agolia. But I am not able to form the hierarchical menu on the frontend
## My data set is
![screen shot 2018-07-13 at 6 28 08 pm](https://user-images.githubusercontent.com/18291428/42692834-9a0b4ee4-86ca-11e8-9b1b-cc0f3eafc528.png)

## Config 
![screen shot 2018-07-13 at 6 28 19 pm](https://user-images.githubusercontent.com/18291428/42692841-a180eada-86ca-11e8-97c4-2d637b3ac4ee.png)

**Comment by bobylito** [comment]
> One problem I am facing here. I have indexed the category details in agolia. But I am not able to form the hierarchical menu on the frontend

@sandipbhuyan Is your problem related to the checkbox issue? Or is it a general enquiry on how to user the hierarchical menu?

**Comment by jasperkennis** [comment]
I've written a little custom version of the menu that could probably be extended into this functionality. Dropping it here now, might pick it up further myself later.

https://gist.github.com/jasperkennis/977b2638068082800a473473d6d48222

**Comment by sjelfull** [comment]
You have my vote as well.

Is there any workarounds for the original request: multi-select with checkboxes?

**Comment by jasperkennis** [comment]
I don't see a solution that plays nice with the intended usage of instantsearch, or even the actual search logic of Algolia. 

So far, the only solution I can come up with is the following (and I don't feel like programming this...). I think you could clone your entire instantseach intance when a user wants to filter for more then one hierarchical attribute, so you end up with several queries ran paralel. The results should then be combined through some kind of merge, rendering the result with some abstraction of the "normal" widgets.
